File: Vb/ImageProcessing/Lib/ComCharts.vb
Imports System.Drawing.Imaging
Imports System.Runtime.InteropServices.Marshal
Imports System.Text

''' <summary>
''' チャートのロジック
''' </summary>
Public MustInherit Class ComCharts
    Protected mHistgram(ComInfo.PictureType.MAX - 1, ComInfo.RGB_MAX - 1) As Integer
    Protected mBitmapOrg As Bitmap
    Protected mBitmapAfter As Bitmap

    ''' <summary>
    ''' コンストラクタ
    ''' </summary>
    Public Sub New()
    End Sub

    ''' <summary>
    ''' デスクトラクタ
    ''' </summary>
    Protected Overrides Sub Finalize()
        MyBase.Finalize()
    End Sub

    ''' <summary>
    ''' イメージからヒストグラム用のデータ算出
    ''' </summary>
    Public Sub CalHistgram()
        Dim widthSize = mBitmapOrg.Width
        Dim heightSize = mBitmapOrg.Height

        Dim bitmapDataOrg = mBitmapOrg.LockBits(New Rectangle(0, 0, widthSize, heightSize), ImageLockMode.ReadWrite, PixelFormat.Format32bppArgb)
        Dim bitmapDataAfter As BitmapData = Nothing
        If (mBitmapAfter IsNot Nothing) Then
            bitmapDataAfter = mBitmapAfter.LockBits(New Rectangle(0, 0, widthSize, heightSize), ImageLockMode.ReadOnly, PixelFormat.Format32bppArgb)
        End If

        Dim indexWidth As Integer
        Dim indexHeight As Integer

        For indexHeight = 0 To heightSize - 1 Step 1
            For indexWidth = 0 To widthSize - 1 Step 1
                Dim position = indexHeight * bitmapDataOrg.Stride + indexWidth * 4
                Dim pixelB = ReadByte(bitmapDataOrg.Scan0, position + ComInfo.Pixel.B)
                Dim pixelG = ReadByte(bitmapDataOrg.Scan0, position + ComInfo.Pixel.G)
                Dim pixelR = ReadByte(bitmapDataOrg.Scan0, position + ComInfo.Pixel.R)

                Dim grayScale As Integer = (pixelB + pixelG + pixelR) / 3

                mHistgram(ComInfo.PictureType.Original, grayScale) += 1

                If (mBitmapAfter IsNot Nothing) Then
                    position = indexHeight * bitmapDataAfter.Stride + indexWidth * 4

                    pixelB = ReadByte(bitmapDataAfter.Scan0, position + ComInfo.Pixel.B)
                    pixelG = ReadByte(bitmapDataAfter.Scan0, position + ComInfo.Pixel.G)
                    pixelR = ReadByte(bitmapDataAfter.Scan0, position + ComInfo.Pixel.R)

                    grayScale = (pixelB + pixelG + pixelR) / 3

                    mHistgram(ComInfo.PictureType.After, grayScale) += 1
                End If
            Next
        Next
    End Sub

    ''' <summary>
    ''' ヒストグラム用のデータ初期化
    ''' </summary>
    Public Sub InitHistgram()
        For index As Integer = 0 To (mHistgram.Length >> 1) - 1 Step 1
            mHistgram(ComInfo.PictureType.Original, index) = 0
            mHistgram(ComInfo.PictureType.After, index) = 0
        Next
    End Sub

    ''' <summary>
    ''' ヒストグラム用のデータCSV保存
    ''' </summary>
    ''' <returns>CSV保存の結果 成功/失敗</returns>
    Public Function SaveCsv() As Boolean
        Dim result = True
        Dim saveDialog = New ComSaveFileDialog()
        saveDialog.Filter = "CSV|*.csv"
        saveDialog.Title = "Save the csv file"
        saveDialog.FileName = "default.csv"
        If (saveDialog.ShowDialog() = True) Then
            Dim delmiter = ","
            Dim stringBuilder = New StringBuilder()
            For nIdx As Integer = 0 To (mHistgram.Length >> 1) - 1
                stringBuilder.Append(nIdx).Append(delmiter)
                stringBuilder.Append(mHistgram(0, nIdx)).Append(delmiter)
                stringBuilder.Append(mHistgram(1, nIdx)).Append(delmiter)
                stringBuilder.Append(Environment.NewLine)
            Next
            If (saveDialog.SreamWrite(stringBuilder.ToString()) = False) Then
                result = False
            End If
        End If

        Return result
    End Function
End Class

File: CSharp/ImageProcessing/Lib/ComCharts.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Drawing.Imaging;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Media.Imaging;

/// <summary>
/// チャートのロジック
/// </summary>
abstract class ComCharts
{
    protected int[,] mHistgram;
    protected Bitmap mBitmapOrg;
    protected Bitmap mBitmapAfter;

    /// <summary>
    /// コンストラクタ
    /// </summary>
    public ComCharts()
    {
        mHistgram = new int[(int)ComInfo.PictureType.MAX, ComInfo.RGB_MAX];
    }

    /// <summary>
    /// デスクトラクタ
    /// </summary>
    ~ComCharts()
    {
    }

    /// <summary>
    /// イメージからヒストグラム用のデータ算出
    /// </summary>
    public void CalHistgram()
    {
        int widthSize = mBitmapOrg.Width;
        int heightSize = mBitmapOrg.Height;

        BitmapData bitmapDataOrg = mBitmapOrg.LockBits(new Rectangle(0, 0, widthSize, heightSize), ImageLockMode.ReadWrite, PixelFormat.Format32bppArgb);
        BitmapData bitmapDataAfter = null;
        if (mBitmapAfter != null)
        {
            bitmapDataAfter = mBitmapAfter.LockBits(new Rectangle(0, 0, widthSize, heightSize), ImageLockMode.ReadWrite, PixelFormat.Format32bppArgb);
        }

        int indexWidth;
        int indexHeight;

        unsafe
        {
            for (indexHeight = 0; indexHeight < heightSize; indexHeight++)
            {
                for (indexWidth = 0; indexWidth < widthSize; indexWidth++)
                {
                    byte* pixel = (byte*)bitmapDataOrg.Scan0 + indexHeight * bitmapDataOrg.Stride + indexWidth * 4;
                    byte grayScale = (byte)((pixel[(int)ComInfo.Pixel.B] + pixel[(int)ComInfo.Pixel.G] + pixel[(int)ComInfo.Pixel.R]) / 3);

                    mHistgram[(int)ComInfo.PictureType.Original, grayScale] += 1;

                    if (mBitmapAfter != null)
                    {
                        pixel = (byte*)bitmapDataAfter.Scan0 + indexHeight * bitmapDataAfter.Stride + indexWidth * 4;
                        grayScale = (byte)((pixel[(int)ComInfo.Pixel.B] + pixel[(int)ComInfo.Pixel.G] + pixel[(int)ComInfo.Pixel.R]) / 3);

                        mHistgram[(int)ComInfo.PictureType.After, grayScale] += 1;
                    }
                }
            }
            mBitmapOrg.UnlockBits(bitmapDataOrg);
            if (mBitmapAfter != null)
            {
                mBitmapAfter.UnlockBits(bitmapDataAfter);
            }
        }
    }

    /// <summary>
    /// ヒストグラム用のデータ初期化
    /// </summary>
    public void InitHistgram()
    {
        for (int index = 0; index < (mHistgram.Length >> 1); index++)
        {
            mHistgram[(int)ComInfo.PictureType.Original, index] = 0;
            mHistgram[(int)ComInfo.PictureType.After, index] = 0;
        }
    }

    /// <summary>
    /// ヒストグラム用のデータCSV保存
    /// </summary>
    /// <returns>CSV保存の結果 成功/失敗</returns>
    public bool SaveCsv()
    {
        bool result = true;
        var saveDialog = new ComSaveFileDialog
        {
            Filter = "CSV|*.csv",
            Title = "Save the csv file",
            FileName = "default.csv"
        };
        if (saveDialog.ShowDialog() == true)
        {
            string delmiter = ",";
            StringBuilder stringBuilder = new StringBuilder();
            for (int index = 0; index < (mHistgram.Length >> 1); index++)
            {
                stringBuilder.Append(index).Append(delmiter);
                stringBuilder.Append(mHistgram[(int)ComInfo.PictureType.Original, index]).Append(delmiter);
                stringBuilder.Append(mHistgram[(int)ComInfo.PictureType.After, index]).Append(delmiter);
                stringBuilder.Append(Environment.NewLine);
            }
            if (!saveDialog.StreamWrite(stringBuilder.ToString()))
            {
                result = false;
            }
        }

        return result;
    }
}
